cmd/ingester/app/consumer/consumer_test.go

```diff
@@ -15,157 +15,187 @@
+	mc := saramaConsumer.ExpectConsumePartition(topic, partition, msgOffset)
+	mc.ExpectMessagesDrainedOnClose()
+
+	saramaPartitionConsumer, e := saramaConsumer.ConsumePartition(topic, partition, msgOffset)
+	require.NoError(t, e)
+
+	undertest := newConsumer(localFactory, topic, mp, newSaramaClusterConsumer(saramaPartitionConsumer))
+
+	undertest.partitionIDToState = map[int32]*consumerState{
+		partition: {
+			partitionConsumer: &partitionConsumerWrapper{
+				topic:             topic,
+				partition:         partition,
+				PartitionConsumer: &kmocks.PartitionConsumer{},
+			},
+		},
+	}
+
+	undertest.Start()
+
+	mc.YieldMessage(msg)
+	isProcessed.Wait()
+
+	mp.AssertExpectations(t)
+	// Ensure that the partition consumer was updated in the map
+	assert.Equal(t, saramaPartitionConsumer.HighWaterMarkOffset(),
+		undertest.partitionIDToState[partition].partitionConsumer.HighWaterMarkOffset())
+	undertest.Close()
+
+	partitionTag := map[string]string{"partition": fmt.Sprint(partition)}
+	testutils.AssertCounterMetrics(t, localFactory, testutils.ExpectedMetric{
+		Name:  "sarama-consumer.messages",
+		Tags:  partitionTag,
+		Value: 1,
+	})
+	testutils.AssertGaugeMetrics(t, localFactory, testutils.ExpectedMetric{
+		Name:  "sarama-consumer.current-offset",
+		Tags:  partitionTag,
+		Value: 1,
+	})
+	testutils.AssertGaugeMetrics(t, localFactory, testutils.ExpectedMetric{
+		Name:  "sarama-consumer.offset-lag",
+		Tags:  partitionTag,
+		Value: 0,
 	})
 }
 
 func TestSaramaConsumerWrapper_start_Errors(t *testing.T) {
-	withWrappedConsumer(func(c *consumerTest) {
-		errCh := make(chan *sarama.ConsumerError, 1)
-		errCh <- &sarama.ConsumerError{
-			Topic: "some-topic",
-			Err:   errors.New("some error"),
-		}
-
-		msgCh := make(chan *sarama.ConsumerMessage)
-
-		c.partitionConsumer.On("Partition").Return(int32(0))
-		c.partitionConsumer.On("Errors").Return((<-chan *sarama.ConsumerError)(errCh))
-		c.partitionConsumer.On("Messages").Return((<-chan *sarama.ConsumerMessage)(msgCh))
-		c.partitionConsumer.On("Close").Return(nil)
-
-		c.consumer.Start()
-		time.Sleep(100 * time.Millisecond)
-		close(msgCh)
-		close(errCh)
-		c.consumer.Close()
-		f := (c.consumer.metricsFactory).(*metrics.LocalFactory)
-		partitionTag := map[string]string{"partition": "0"}
-		testutils.AssertCounterMetrics(t, f, testutils.ExpectedMetric{
-			Name:  "sarama-consumer.errors",
-			Tags:  partitionTag,
-			Value: 1,
-		})
+	localFactory := metrics.NewLocalFactory(0)
+
+	saramaConsumer := smocks.NewConsumer(t, &sarama.Config{})
+	mc := saramaConsumer.ExpectConsumePartition(topic, partition, msgOffset)
+	mc.ExpectErrorsDrainedOnClose()
+
+	saramaPartitionConsumer, e := saramaConsumer.ConsumePartition(topic, partition, msgOffset)
+	require.NoError(t, e)
+
+	undertest := newConsumer(localFactory, topic, &pmocks.SpanProcessor{}, newSaramaClusterConsumer(saramaPartitionConsumer))
+
+	undertest.Start()
+	mc.YieldError(errors.New("Daisy, Daisy"))
+	time.Sleep(100 * time.Millisecond)
```

Comment:
any other condition we can wait on?

Reply:
We can wait on the metric changing, but that would involve a sleep/poll as well. Processing errors only prints out a log and increments a counter, so not sure we can eliminate this sleep entirely in tests.

Reply:
can we keep checking the metric to see if it's incremented?

Reply:
We can poll the metric assertion until it's true, but do we need to? (I don't think we do such polling on other tests)